modules/sync/unique_queue.go

```diff
@@ -82,6 +82,7 @@ func (q *UniqueQueue) AddFunc(id interface{}, fn func()) {
 	idStr := com.ToStr(id)
 	q.table.lock.Lock()
 	if _, ok := q.table.pool[idStr]; ok {
+		q.table.lock.Unlock()
```

Comment:
Can use defer instead?

Reply:
nope you want to unlock before you push to the channel.

Reply:
the only issue is if you panic in fn() then you might not get unlocked...

Reply:
really not sure what you can do to protect against that.

Reply:
I suppose this would cope:

```go
func (q *UniqueQueue) AddFunc(id interface{}, fn func()) {
	idStr := com.ToStr(id)
	q.table.lock.Lock()
	locked := true
	defer func() {
		if locked {
			q.table.lock.Unlock()
		}
	}()
	if _, ok := q.table.pool[idStr]; ok {
		return
	}
	q.table.pool[idStr] = struct{}{}
	if fn != nil {
		fn()
	}
	locked = false
	q.table.lock.Unlock()
	select {
	case <-q.closed:
		return
	case q.queue <- idStr:
		return
	}
}
```